scripts/git/pre-commit

```diff
@@ -0,0 +1,20 @@
+#!/bin/sh
+#
+# This script ensures Crystal code is correctly formatted before committing it.
+# It won't apply any format changes automatically.
+#
+# Only staged files (the ones to be committed) are being processed, but each file is checked
+# entirely as it is stored on disc, even parts that are not staged.
+#
+# To use this script, it needs to be installed in the local git repository. For example by running
+# `ln -s scripts/git/pre-commit .git/hooks` in the root folder.
+#
+# Called by "git commit" with no arguments. The hook should
+# exit with non-zero status after issuing an appropriate message if
+# it wants to stop the commit.
+
+changed_cr_files=$(git diff --cached --name-only --diff-filter=A | grep '\.cr$')
+
+[ -z "$changed_cr_files" ] && exit 0
+
+exec crystal tool format --check "$changed_cr_files" >&2
```

Comment:
`"$changed_cr_files"` is wrong, `$chnaged_cr_files` (without wrapping quote) is correct. `$changed_cr_files` is multiline text when there are two or more changed files, and bare variable expands each lines as one command arg.

Reply:
@MakeNowJust but that won't work if there is a space in the path to one of the file..
A workaround would be to make a shell array, and expand it with `"${the_array[@]}"` which is a shell trick to generate quotes arguments for each item of `the_array`

Reply:
@bew in anyway we have to avoid Bashisms (no arrays). Some of us are using Alpine/BSD

Reply:
How do you handle path with space then?

Reply:
`git diff --cached --name-only --diff-filter=ACM | awk '/^.*\.cr$/ { gsub(" ", "\ "); print $0 }'` do the trick - spaces need to be escaped on UNIX

Reply:
It is hard problem. Escaping space by hand is not so hard. `sed` is useful, but, we should escape any other **shell special characters** even if space is escaped. It is hard.  Additionally, filename may have newline, it breaks one-line-one-filename rule. We can use `-z` option of `git diff`, but handling null character on POSIX shell is so hard (almost impossible). Otherwise, `git diff --name-only ...` output is changed by `core.quotepath` and some git configurations, so we should set correct values if consistent results are needed.

In the real world, when you add a file whose name contains spaces to this repository? I never do, and I believe you never too.

It is not valuable that we mind this problem. In fact Go's `pre-commit` hook script does not pay attention for it.

It is terrible when CI script is broken. But, when `pre-commit` script is broken, we can skip it via `-n` option.

This `"$changed_cr_files"` is **absolutely broken**. It must not work against multiple staged files. It should be fixed.

Reply:
Of course we should add escaping space trick if we add path with space file in near future.

Reply:
Agree @MakeNowJust . In crystal/crystal-lang, we'll never have files containing spaces (or I hope, because there is no point to).
I the project absolutely need a space in a file name, a trick could be reconsidered.

Reply:
I don't think we need to care about pathnames including spaces or special characters. Yes, there are possible flaws but it's a non-trivial problem (sadly).

I'm sure such paths wouldn't be approved in the Crystal repo anyway. So there is no need to bother validating them in a pre-commit hook.

Reply:
@straight-shoota It has trivial problem for now. You should fix.